include/alpaka/algo/Transform.hpp

```diff
@@ -0,0 +1,192 @@
+            {
+                static_assert(std::is_invocable_r_v<T, TFn, T> or std::is_invocable_r_v<T, TFn, TAcc const&, T>);
+
+                using Dim = alpaka::Dim<TAcc>;
+                using Idx = alpaka::Idx<TAcc>;
+                using Vec = alpaka::Vec<Dim, Idx>;
+
+                for(Vec idx : alpaka::uniformElementsND(acc, in_size))
+                {
+                    auto p_in
+                        = reinterpret_cast<T const*>(reinterpret_cast<uintptr_t>(in_ptr) + (idx * in_pithces).sum());
+                    auto p_out
+                        = reinterpret_cast<T*>(reinterpret_cast<uintptr_t>(out_ptr) + (idx * out_pitches).sum());
+                    if constexpr(std::is_invocable_r_v<T, TFn, T>)
+                    {
+                        // std::is_invocable_r_v<T, TFn, T>
+                        *p_out = fn(*p_in);
+                    }
+                    else
+                    {
+                        // std::is_invocable_r_v<T, TFn, TAcc const&, T>
+                        *p_out = fn(acc, *p_in);
+                    }
+                }
+            }
+        };
+
+    } // namespace detail
+
+    /*
+     * Applies asynchronously the given function `fn` to the elements of the semi-open input range
+     * `[first_in, last_in)`, and stores the result in an output range starting from `first_out`,
+     * using the accelerator back-end identified by `Tag`.
+     */
+    template<typename TTag, typename TQueue, typename T, typename TFn>
+    void transform(TQueue& queue, T const* first, T const* last, T* out, TFn&& fn)
+    {
+        using Idx = typename std::iterator_traits<T*>::difference_type;
+        using Acc1D = alpaka::TagToAcc<TTag, alpaka::DimInt<1>, Idx>;
+
+        static_assert(
+            std::is_invocable_r_v<T, TFn, T> or std::is_invocable_r_v<T, TFn, Acc1D const&, T>,
+            "TFn must accept either one argument (of type T) or two arguments (an accelerator and an argument of type "
+            "T), and return a value of type T.");
+
+        Idx size = std::distance(first, last);
+        detail::TransformKernel<TFn> kernel{fn};
+
+        // Find a valid work division. This could be further optimised.
+        auto const config
+            = alpaka::KernelCfg<Acc1D>{size, Idx{1}, false, alpaka::GridBlockExtentSubDivRestrictions::Unrestricted};
+        auto const grid = alpaka::getValidWorkDiv(config, alpaka::getDev(queue), kernel, first, out, size);
+
+        // Apply the fn function to all elements of the input range.
+        alpaka::exec<Acc1D>(queue, grid, kernel, first, out, size);
+    }
+
+    /*
+     * Applies asynchronously the given function `fn` to the elements of the input buffer `in`,
+     * and stores the result in the corresponding elements of the output buffer `out`,
+     * using the accelerator back-end identified by `Tag`.
+     */
+    template<typename TTag, typename TQueue, typename TConstBuf, typename TBuf, typename TFn>
```

Comment:
```suggestion
    template<alpaka::concepts::Tag TTag, typename TQueue, typename TConstBuf, typename TBuf, typename TFn>
```

A concept for `alpaka::Buf` and `alpaka::ConstBuf` would be also nice, but does not exist. I think this is something for an upcoming PR.